File: src/main/java/com/google/cloud/spanner/pgadapter/commands/Command.java
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.cloud.spanner.pgadapter.commands;

import com.google.api.core.InternalApi;
import com.google.cloud.spanner.connection.Connection;
import com.google.cloud.spanner.pgadapter.metadata.DynamicCommandMetadata;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.json.simple.JSONObject;

/**
 * This abstract class concerns itself with representing a method of matching a specific statement
 * and translating it to a form that Spanner understands. The user should call is() to determine
 * whether it matches before trying to translate.
 */
@InternalApi
public abstract class Command {

  protected Matcher matcher;
  protected String sql;

  Command(String sql) {
    this.sql = sql;
    this.matcher = getPattern().matcher(sql);
  }

  /**
   * This constructor should only be called if you do wish to do your own matching (i.e.: set up
   * this.matcher). Useful if the pattern is not static, or not pattern related (in which case you
   * may set it to null).
   */
  Command(Matcher matcher) {
    this.matcher = matcher;
  }

  /**
   * The list of all subcommands to this command. When subclassing this, make sure to add the new
   * command to the below list, if applicable (i.e.: if you want it to be run). Commands may also be
   * specified via Dynamic Commands, which are defined from an input JSON file. These commands
   * precede the hard-coded ones in execution.
   *
   * @param sql The SQL command to be translated.
   * @param connection The connection currently in use.
   * @return A list of all Command subclasses.
   */
  public static List<Command> getCommands(
      String sql, Connection connection, JSONObject commandMetadataJSON) {
    List<Command> commandList = new ArrayList<>();

    for (DynamicCommandMetadata currentMetadata :
        DynamicCommandMetadata.fromJSON(commandMetadataJSON)) {
      commandList.add(new DynamicCommand(sql, currentMetadata));
    }

    commandList.addAll(
        Arrays.asList(new ListCommand(sql, connection), new InvalidMetaCommand(sql)));

    return commandList;
  }

  /**
   * The pattern expected by this command: is() will only return true if it matches this pattern.
   *
   * @return The pattern apposite (in relation to) the command.
   */
  protected abstract Pattern getPattern();

  /** @return True if the SQL statement matches the pattern. */
  public boolean is() {
    return this.matcher.find();
  }

  /** @return The equivalent Spanner statement for the specified SQL statement. */
  public abstract String translate();
}

File: src/rust/spanner-pgadapter/src/pgwire/command.rs
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// Extracts the SQL command tag (e.g. SELECT, INSERT, UPDATE, DELETE) from the SQL string.
/// TODO: This method should be moved into the Rust Connection API parser.
pub fn parse_command_tag(sql: &str) -> String {
    let mut clean_sql = String::new();
    let mut characters = sql.chars().peekable();

    while let Some(character) = characters.next() {
        if character == '-' && characters.peek() == Some(&'-') {
            // Skip line comments
            while let Some(next_char) = characters.next() {
                if next_char == '\n' || next_char == '\r' {
                    break;
                }
            }
        } else if character == '/' && characters.peek() == Some(&'*') {
            // Skip block comments
            characters.next(); // consume '*'
            while let Some(next_char) = characters.next() {
                if next_char == '*' && characters.peek() == Some(&'/') {
                    characters.next(); // consume '/'
                    break;
                }
            }
        } else {
            clean_sql.push(character);
        }
    }

    if let Some(first_word) = clean_sql.trim().split_whitespace().next() {
        first_word.to_uppercase()
    } else {
        "SELECT".to_string()
    }
}

/// Formats a PG CommandComplete tag based on the raw command name (e.g. INSERT, SELECT) and row/affected count.
pub fn format_command_complete_tag(command: &str, count: usize) -> String {
    if command == "INSERT" {
        format!("INSERT 0 {}", count)
    } else {
        format!("{} {}", command, count)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_command_tag() {
        assert_eq!(parse_command_tag("SELECT 1"), "SELECT");
        assert_eq!(parse_command_tag("  insert into foo values (1)"), "INSERT");
        assert_eq!(
            parse_command_tag("-- comment\nUPDATE bar SET x = 1"),
            "UPDATE"
        );
        assert_eq!(
            parse_command_tag("/* block comment */delete from baz"),
            "DELETE"
        );
        assert_eq!(parse_command_tag(""), "SELECT");
    }

    #[test]
    fn test_format_command_complete_tag() {
        assert_eq!(format_command_complete_tag("SELECT", 5), "SELECT 5");
        assert_eq!(format_command_complete_tag("INSERT", 12), "INSERT 0 12");
        assert_eq!(format_command_complete_tag("UPDATE", 0), "UPDATE 0");
        assert_eq!(format_command_complete_tag("DELETE", 3), "DELETE 3");
    }
}
